tinygrad/runtime/ops_qcom.py

```diff
@@ -0,0 +1,291 @@
+    if value is not None: self._patch(cmd_idx, offset=3, data=data64_le(value))
+
+  def _update_wait(self, cmd_idx, signal, value):
+    if signal is not None: self._patch(cmd_idx, offset=2, data=data64_le(mv_address(signal._signal)))
+    if value is not None: self._patch(cmd_idx, offset=4, data=[value & 0xffffffff])
+
+  def _timestamp(self, signal):
+    # doesnt really work right now
+    pass
+    self.cmd(adreno.CP_REG_TO_MEM, 0x40000980,  *data64_le(mv_address(signal._signal)))
+    self.cmd(adreno.CP_EVENT_WRITE7, adreno.CACHE_INVALIDATE)
+
+  def bind(self, device):
+    self.binded_device = device
+
+    cmdbytes = array.array('I', self.q)
+    self.hw_page = device._gpu_alloc(len(cmdbytes) * 4, 0xC0A00, map_to_cpu=True)
+    ctypes.memmove(self.hw_page.va_addr, mv_address(memoryview(cmdbytes)), len(cmdbytes) * 4)
+
+    self.obj = kgsl.struct_kgsl_command_object(gpuaddr=self.hw_page.va_addr, size=len(cmdbytes) * 4, flags=0x1)
+    self.submit_req = kgsl.struct_kgsl_gpu_command(cmdlist=ctypes.addressof(self.obj), cmdsize=ctypes.sizeof(kgsl.struct_kgsl_command_object),
+                                                   numcmds=1, context_id=device.ctx)
+    self.q = to_mv(self.hw_page.va_addr, self.hw_page.size).cast("I")
+
+  def _submit(self, device: QcomDevice):
+    if self.binded_device == device:
+      device._ioctl(kgsl.IOCTL_KGSL_GPU_COMMAND, self.submit_req)
+      return
+
+    cmdbytes = array.array('I', self.q)
+    alloc = device._gpu_alloc(len(cmdbytes) * 4, 0xC0A00, map_to_cpu=True)
+    ctypes.memmove(alloc.va_addr, mv_address(memoryview(cmdbytes)), len(cmdbytes) * 4)
+
+    obj = kgsl.struct_kgsl_command_object(gpuaddr=alloc.va_addr, size=len(cmdbytes) * 4, flags=0x1)
+    submit_req = kgsl.struct_kgsl_gpu_command(cmdlist=ctypes.addressof(obj), cmdsize=ctypes.sizeof(kgsl.struct_kgsl_command_object),
+                                              numcmds=1, context_id=device.ctx)
+    device._ioctl(kgsl.IOCTL_KGSL_GPU_COMMAND, submit_req)
+
+  def _exec(self, prg, kernargs, global_size, local_size):
+    global_size_mp = cast(Tuple[int,int,int], tuple(int(g*l) for g,l in zip(global_size, local_size))) if local_size else global_size
+
+    self.cmd_idx_to_dims[len(self) - 1] = [global_size, local_size]
+    self.reg(
+      adreno.REG_A6XX_HLSQ_CS_NDRANGE_0,
+        # kernel dimenstion = 3
+        3 | ((local_size[0] - 1) << adreno.A6XX_HLSQ_CS_NDRANGE_0_LOCALSIZEX__SHIFT)
+        | ((local_size[1] - 1) << adreno.A6XX_HLSQ_CS_NDRANGE_0_LOCALSIZEY__SHIFT)
+        | ((local_size[2] - 1) << adreno.A6XX_HLSQ_CS_NDRANGE_0_LOCALSIZEZ__SHIFT),
+        global_size_mp[0], 0, global_size_mp[1], 0, global_size_mp[2], 0, # global size x,y,z followed by offsets
+        0xccc0cf, 0x2fc,
+        global_size[0], global_size[1], global_size[2], # original global sizes
+    )
+    self.cmd(adreno.CP_SET_MARKER, adreno.RM6_COMPUTE)
```

Comment:
can we move this into device's `__init__`?

Reply:
done